src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
-     */
-    ByteBuffer overlapDetection(ByteBuffer src, ByteBuffer dst) {
-        if (src.isDirect() && dst.isDirect()) {
-            DirectBuffer dsrc = (DirectBuffer) src;
-            DirectBuffer ddst = (DirectBuffer) dst;
+        /**
+         * Use any data from ibuffer and 'src' to first verify the auth tag. If
+         * the tag is valid, decrypt the data.
+         */
+        @Override
+        public int doFinal(ByteBuffer src, ByteBuffer dst)
+            throws IllegalBlockSizeException, AEADBadTagException,
+            ShortBufferException {
+            GHASH save = null;
+
+            // If these are array backed ByteBuffers, use the underlying array
+            if (src.hasArray() && dst.hasArray()) {
+                int len = doFinal(src.array(),
+                    src.arrayOffset() + src.position(),
+                    src.remaining(), dst.array(),
+                    dst.arrayOffset() + dst.position());
+                src.position(src.limit());
+                dst.position(dst.position() + len);
+                return len;
+            }
+
+            // Check for overlap in the bytebuffers
+            dst = overlapDetection(src, dst);
 
-            // Get the current memory address for the given ByteBuffers
-            long srcaddr = dsrc.address();
-            long dstaddr = ddst.address();
+            // Length of the input
+            ByteBuffer tag;
+            ByteBuffer ct = src.duplicate();
 
-            // Find the lowest attachment that is the base memory address of the
-            // shared memory for the src object
-            while (dsrc.attachment() != null) {
-                srcaddr = ((DirectBuffer) dsrc.attachment()).address();
-                dsrc = (DirectBuffer) dsrc.attachment();
+            ByteBuffer buffer = ((ibuffer == null || ibuffer.size() == 0) ?
+                null : ByteBuffer.wrap(ibuffer.toByteArray()));
+            int len;
+
+            if (ct.remaining() >= tagLenBytes) {
+                tag = src.duplicate();
+                tag.position(ct.limit() - tagLenBytes);
+                ct.limit(ct.limit() - tagLenBytes);
+                len = ct.remaining();
+                if (buffer != null) {
+                    len += buffer.remaining();
+                }
+            } else if (buffer != null) {
+                // It's unlikely the tag will be between the buffer and data
+                tag = ByteBuffer.allocate(tagLenBytes);
+                int limit = buffer.remaining() - (tagLenBytes - ct.remaining());
+                buffer.mark();
+                buffer.position(limit);
+                // Read from "new" limit to buffer's end
+                tag.put(buffer);
+                // reset buffer to data only
+                buffer.reset();
+                buffer.limit(limit);
+                tag.put(ct);
+                tag.flip();
+                // Limit is how much of the ibuffer has been chopped off.
```

Comment:
I think it's which direction you are looking at it from, when I wrote this I was thinking how much of the ibuffer was chopped off into tag.. i can reword that because its ambiguous.